### PR TITLE
Add Adapter Selection Sample

### PR DESCRIPTION
**Pull request by nums11**
This change adds the adapter selection sample to the Windows ML Samples Gallery.
- An AdapterList class was added to the WinMLSamplesGalleryNative project to handle adapter selection using DXCore
- In order to use DXCore, the target sdk version was updated from 10.0.18362.0 to 10.0.19041.0

### CODE REVIEW

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                </ListView.ItemContainerStyle>
+            </ListView>
+        </Border>
+        <ComboBox x:Name="CodeSnippetComboBox" SelectedIndex="0" BorderBrush="#dbdbdb" Header="" Width="70" FontSize="18" Margin="0,20,0,0"
+                  Typography.Capitals="AllSmallCaps" Typography.StylisticSet4="True"
+                  HorizontalAlignment="Right">
+            <x:String>C#</x:String>
+        </ComboBox>
+        <Grid x:Name="CodeSnippet" Background="#F6F6F6" BorderBrush="#dbdbdb" BorderThickness="1" CornerRadius="5" Padding="15,20,15,22">
+            <TextBlock x:Name="TestName" FontSize="14" FontFamily="Consolas">
```
Comment (linnealovespie) [inline Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml:35]: Just a comment but you can render markdown text in XAML, which you can then use to write code snippets in various languages. https://docs.microsoft.com/en-us/windows/communitytoolkit/controls/markdowntextblock

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json [hunk 2]
```diff
@@ -71,6 +71,17 @@
       "CSharpGithubLink": "https://github.com/microsoft/Windows-Machine-Learning/blob/master/Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/EncryptedModel/EncryptedModel.xaml.cs",
       "Docs": [],
       "IsRecentlyAdded": true
+    },
+    {
+      "Title": "Adapter Selection",
+      "DescriptionShort": "Select different adapters based on your power and performance needs.",
+      "Description": "The sample showcases how to use Windows ML with different adapters that have tradeoffs between power and performance.",
```
Comment (linnealovespie) [inline Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json:78]: I didn't know adapters = a gpu or cpu device, could the description have a bit of an explanation on adapters?

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml (hunk 1 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml:35]: Is this only for UWP apps? I can't seem to find the markdowntextblock in the Microsoft.UI.Xaml namespace

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json (hunk 2 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json:78]: I'm honestly still trying to understand what exactly constitutes an adapter myself. The answers that I've seen online seem a little vague so I'd be open to suggestions from anyone as to what a good description might be.

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Module.ixx [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export module Module;
+
+export void MyFunc();
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Module.ixx:3]: Why is this file needed?

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+        // Retrieve an adapter list
+        winrt::com_ptr<IDXCoreAdapterList> d3D12CoreComputeAdapters;
+        GUID attributes[]{ DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
+        winrt::check_hresult(
+            adapterFactory->CreateAdapterList(_countof(attributes),
+                attributes,
+                d3D12CoreComputeAdapters.put()));
+
+        // Find the driver in the list with the matching description
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp:54]: I think should be adapter instead of driver in the comment

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                d3D12CoreComputeAdapters->GetAdapter(i, candidateAdapter.put()));
+            CHAR description[128];
+            candidateAdapter->GetProperty(DXCoreAdapterProperty::DriverDescription, sizeof(description), description);
+            driver_descriptions[i] = to_hstring(description);
+        }
+
+        return driver_descriptions;
+    }
+
+    winrt::Microsoft::AI::MachineLearning::LearningModelDevice AdapterList::CreateLearningModelDeviceFromAdapter(winrt::hstring description) {
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp:41]: Seems like there is repeated code between this function and GetAdapters(), do you think the code can be shared?

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Rules in this file were initially inferred by Visual Studio IntelliCode from the C:\Users\numform\Windows-Machine-Learning\Samples\WinMLSamplesGallery\WinMLSamplesGalleryNative\ codebase based on best match to current usage at 12/9/2021
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig:1]: What is this file used for?

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Module.ixx (hunk 3 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/Module.ixx]: Not sure... Seems to be generated but I removed it.

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig (hunk 6 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig:1]: It's used by visual studio to maintain things like indentation for the project

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp (hunk 4 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp]: Fixed.

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp (hunk 5 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp:46]: Done.

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig (hunk 6 above)
Comment (ryanlai2) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig:1]: I don't think this file is needed. There's no C# coding convension rules listed even though [*.cs] is here.

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj.filters [hunk 7]
```diff
@@ -37,6 +43,7 @@
     <None Include="squeezenet1.1-7-batched.onnx" />
     <None Include="WinMLSamplesGalleryNative.def" />
     <None Include="PropertySheet.props" />
+    <None Include=".editorconfig" />
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj.filters:46]: I don't think this file is needed

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json [hunk 8]
```diff
@@ -71,6 +71,17 @@
...
+      "Icon": "\uE155",
+      "Tag": "AdapterSelection",
+      "XAMLGithubLink": "https://github.com/microsoft/Windows-Machine-Learning/blob/user/numform/adapter_selection/Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml",
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json:81]: There are various areas in the PR where the link contains a link to your user branch, can we fix it so that it points to master branch?

The files don't exist on master branch yet.. so the links won't work yet

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        {
+            winrt::com_ptr<IDXCoreAdapter> candidateAdapter;
+            winrt::check_hresult(
+                d3D12CoreComputeAdapters->GetAdapter(i, candidateAdapter.put()));
+            CHAR driver_description[128];
+            candidateAdapter->GetProperty(DXCoreAdapterProperty::DriverDescription, sizeof(driver_description), driver_description);
+            hstring hstr_driver_description = to_hstring(driver_description);
+
+            if (description == hstr_driver_description) {
+                found_driver = 1;
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp:62]: Seems like found_driver isn't used, also how come int instead of bool?

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig (hunk 6 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/.editorconfig:1]: Removed.

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj.filters (hunk 7 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj.filters:46]: Removed.

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json (hunk 8 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json:81]: Done.

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp (hunk 9 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/AdapterList.cpp:62]: Removed.

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json (hunk 2 above)
Comment (martinb35) [reply Samples/WinMLSamplesGallery/WinMLSamplesGallery/SampleMetadata/SampleMetadata.json:78]: could we use "device" instead?

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                device = WinMLSamplesGalleryNative.AdapterList.CreateLearningModelDeviceFromAdapter(device_kind_str);
+                toggleCodeSnippet(false);
+            }
+        }
+
+        private LearningModelDeviceKind GetLearningModelDeviceKind(string device_kind_str)
+        {
+            if (device_kind_str == "CPU")
+            {
+                selectedDeviceKind.Text = "Cpu";
```
Comment (martinb35) [inline Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml.cs:58]: why not just use device_kind_str  instead of "Cpu"?

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj [hunk 11]
```diff
@@ -160,6 +162,7 @@
     <Midl Include="WinMLSamplesGalleryNative.idl" />
   </ItemGroup>
   <ItemGroup>
+    <None Include=".editorconfig" />
```
Comment (ryanlai2) [inline Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj:165]: Need to remove this

File: Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml.cs (hunk 10 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGallery/Samples/AdapterSelection/AdapterSelection.xaml.cs]: Done.

**Review by martinb35 — APPROVED** [review]

File: Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj (hunk 11 above)
Comment (nums11) [reply Samples/WinMLSamplesGallery/WinMLSamplesGalleryNative/WinMLSamplesGalleryNative.vcxproj:165]: Done.

**Review by ryanlai2 — APPROVED** [review]
